### PR TITLE
added feature importance

### PR DESCRIPTION
**Pull request by clementstross**

### CODE REVIEW

File: R/feature_importance_permutation.R [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+#'
+#' data <- data.frame(x1=runif(100, 0, 25), x2=runif(100, 0, 25), x3=runif(100, 0, 25)) %>%
+#'   mutate(target=x1^2 * 0.01 + x2 + rnorm(n(),sd=5))
+#'
+#' model_gbm <- xgboost(data = as.matrix(data[,which(!(names(data)=="target"))]), label=data[["target"]], nrounds=20, verbose = 0)
+#'
+#'
+#' feature_importance_permutation(xgb.model=model_gbm,
+#'                               factors_df=data[,which(!(names(data)=="target"))],
+#'                               target=data[[target]])
```
Comment (gloverd2) [inline R/feature_importance_permutation.R:34]: target should be specified or in ""

File: R/feature_importance_permutation.R [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+  # checks on inputs
+  checkmate::assert_data_frame(factors_df)
+  checkmate::assert_numeric(target, len = nrow(factors_df))
+  checkmate::assert_integerish(nrounds, len=1)
+  checkmate::assert_integerish(seed, len=1)
+  checkmate::assert_function(metric, args=c("actual", "predicted", "weight", names(metric_kwarg)))
+  checkmate::assert_list(metric_kwarg, unique = TRUE, null.ok = TRUE)
+
+  if (is.null(weight)){weight <- rep(1, nrow(factors_df))}
```
Comment (gloverd2) [inline R/feature_importance_permutation.R:54]: Can somebody else check this too but for me the if statement runs on it's own without the else statement? I think becaue of the indentation.

File: tests/testthat/test_feature_importance_permutation.R [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  testthat::expect_error(feature_importance_permutation(data[,which(!(names(data)=="target"))], xgb.model=model_gbm, target=data[["target"]], nrounds=c(1,1)))
+  testthat::expect_error(feature_importance_permutation(data[,which(!(names(data)=="target"))], xgb.model=model_gbm, target=data[["target"]], nrounds="10"))
+})
+
+
+testthat::test_that("Testing seed - Should error",{
+  # Testing column must be in data
+  testthat::expect_error(feature_importance_permutation(data[,which(!(names(data)=="target"))], xgb.model=model_gbm, target=data[["target"]], seed=c(1,1)))
+  testthat::expect_error(feature_importance_permutation(data[,which(!(names(data)=="target"))], xgb.model=model_gbm, target=data[["target"]], seed="10"))
+})
```
Comment (gloverd2) [inline tests/testthat/test_feature_importance_permutation.R:43]: do we need decimal?

File: R/feature_importance_permutation.R [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+#' @export
+#'
+#' @examples
+#'
+#' data <- data.frame(x1=runif(100, 0, 25), x2=runif(100, 0, 25), x3=runif(100, 0, 25)) %>%
+#'   mutate(target=x1^2 * 0.01 + x2 + rnorm(n(),sd=5))
+#'
+#' model_gbm <- xgboost(data = as.matrix(data[,which(!(names(data)=="target"))]), label=data[["target"]], nrounds=20, verbose = 0)
+#'
+#'
```
Comment (gloverd2) [inline R/feature_importance_permutation.R:31]: the importance_sd column is NA. is this how it should be ? if its all NA should we drop it ?

**Review by gloverd2 — COMMENTED** [review]
early review. does it need to be restricrted to xgboost? think Duncan has something that worked for other models too

File: R/feature_importance_permutation.R (hunk 1 above)
Comment (clementstross) [reply R/feature_importance_permutation.R:34]: agree

File: R/feature_importance_permutation.R (hunk 4 above)
Comment (clementstross) [reply R/feature_importance_permutation.R:31]: If nrounds is equal to 1 then it is NA i think this is probably what we want. I don't really mind though. I'll change the defaults so it happens less

File: R/feature_importance_permutation.R (hunk 2 above)
Comment (clementstross) [reply R/feature_importance_permutation.R:54]: I have cleaned it up.

File: R/feature_importance_permutation.R (hunk 4 above)
Comment (gloverd2) [reply R/feature_importance_permutation.R:31]: no i am happy for it to be this way. have added a short line after importance_sd to clarify

**Comment by gloverd2** [comment]
What's the overall opinion on whether in tests we should clarify the parameter for each argument such that:
i.e function(parameter = user_input) or are we happy with function(user_input).
i find including the parameter is easier for me to read as a reviewer but i can understand things quickly get irrelevant and messy . Discuss

**Review by gloverd2 — APPROVED** [review]
approved with haste as i think it is very useful alongside the metrics. Im interested to know if its been use in practice?

**Comment by clementstross** [comment]
@CRSlade1 Do you want a look at this? if not I'll merge either today or tomorrow

File: R/feature_importance_permutation.R [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+  permuted_data <- data
+  for (ii in 1:length(model.features)){ # for all features
+
+    var_ii <- model.features[ii]
+    original_values_ii <- data[[var_ii]] # save original ordering for factor
+    scores_permuted_ii <- c() # empty vector to contain scores
+
+    for (kk in 1:nrounds){
+      set.seed((seed * ii) + kk) # set seed
```
Comment (CRSlade1) [inline R/feature_importance_permutation.R:96]: In cases where seed is low, you could end up with a large amount of collinearity.
i.e seed is 1, feature 1 goes from 2:11. seed is 1, feature 2 goes from 3:12.
Seed needs to be above length(model.features) for true independence.

Comment (CRSlade1) [reply R/feature_importance_permutation.R:96]: I think this would also force seed to need to be positive. (Although you could just subtract instead if negative if you really wanted to allow that)

File: R/feature_importance_permutation.R [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                                           weight=NULL,
+                                           metric=metric_rmse,
+                                           nrounds=10,
+                                           metric_kwarg=NULL,
+                                           seed=666){
+
+  # checks on inputs
+  checkmate::assert_data_frame(data)
+  checkmate::assert_numeric(actual, len = nrow(data))
+  checkmate::assert_integerish(nrounds, len=1)
```
Comment (CRSlade1) [inline R/feature_importance_permutation.R:59]: Must be positive

**Review by CRSlade1 — CHANGES_REQUESTED** [review]
Have some concerns about handling of seed and having biased permutations but think it's good except from that

File: R/feature_importance_permutation.R [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+#' @param weight vector[Numeric] - exposure for predictions
+#' @param metric function - Of type in admr::metric_ - must have arguments actual, predicted, weight
+#' @param nrounds integer - Number of times to permute each factor
+#' @param metric_kwarg namedlist - additional optional arguments for function \code{metric}
+#' @param seed integer - random seed for permuations
+#'
+#' @return  dataframe with columns
+#'           :col_index - position of factor in \code{data}
+#'           :factor - name of factor in \code{data}
+#'           :importance_mean - importance of factor
```
Comment (mdpead) [inline R/feature_importance_permutation.R:19]: More changes from factor to feature

File: R/feature_importance_permutation.R [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+  if (is.null(weight)){
+    weight <- rep(1, nrow(data))
+  }else{
+    checkmate::assert_numeric(weight, len = nrow(data), lower=0)
+  }
+
+  # extract feature names
+  model.features <- colnames(data)
+
+  # Get average prediction
```
Comment (mdpead) [inline R/feature_importance_permutation.R:73]: This is the optimal / baseline prediction isn't it?

File: R/feature_importance_permutation.R [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+  if(any(class(model)=="xgb.Booster")){
+
+    # create dmatrix for standard data
+    d_mat <- xgboost::xgb.DMatrix(data=data %>% sapply(as.numeric) %>% as.matrix())
+    predicted <- predict(object=model, newdata=d_mat) %>% as.vector()
+  }else{
+    predicted <- predict(object=model, newdata=data) %>% as.vector()
+  }
```
Comment (mdpead) [inline R/feature_importance_permutation.R:81]: Doesn't make a difference, but since the two `predicted <- ...` lines are the same, could pull this out of the if-else statement

File: R/feature_importance_permutation.R [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+  permuted_data <- data
+  for (ii in 1:length(model.features)){ # for all features
+
+    var_ii <- model.features[ii]
+    original_values_ii <- data[[var_ii]] # save original ordering for factor
+    scores_permuted_ii <- c() # empty vector to contain scores
+
+    for (kk in 1:nrounds){
+      set.seed((((seed+length(model.features)) * ii) + kk)) # set seed
```
Comment (mdpead) [inline R/feature_importance_permutation.R:96]: Maybe I'm missing something, but do we need to set a different seed for each loop? If we just set the seed at the beginning of the function, we will still get the same sequence of random numbers each time, and so the output of the function will be determinate.

File: R/feature_importance_permutation.R [hunk 11]
```diff
@@ -0,0 +1,140 @@
+#' feature_importance_permutation
+#'
+#' @description
+#' returns which columns are the most important in the fitted model.
+#' This is done by permuting the inputs and measuring the deterioration of the metric
+#'
+#' @param data dataframe - named columns and contains only factors used in \code{xgb.mode} in the correct order
+#' @param model xgboost model
```
Comment (mdpead) [inline R/feature_importance_permutation.R:8]: Just to check, we're expecting the xgboost model to be a final model with the optimal number of trees, rather than a trained sequence of trees where an optimal tree parameter is passed in?

File: R/feature_importance_permutation.R [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+        predicted_permuted <- predict(object=model, newdata=permuted_d_mat) %>% as.vector()
+
+      }else{
+        predicted_permuted <- predict(object=model, newdata=permuted_data) %>% as.vector()
+      }
+
+      score_permuted <- metric %>% do.call(append(list(actual=actual, predicted=predicted_permuted, weight=weight),
+                                                  metric_kwarg))
+      scores_permuted_ii[kk] <- score_permuted # append to vector of scores
+    }
```
Comment (mdpead) [inline R/feature_importance_permutation.R:115]: Given that this procedure is exactly the same as for the non-permuted data, would it make more sense to pass both the non-permuted data and each permutation through the same function?

File: R/feature_importance_permutation.R [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+
+    if (ii==1){out_df <- out_df_ii}
+    else{out_df <- rbind(out_df, out_df_ii)}
+
+  }
+
+  # If a low score is good switch the ordering
+  if (out_df[["importance_mean"]] %>% mean() < 0){
+    out_df[["importance_mean"]] = -out_df[["importance_mean"]]
+  }
```
Comment (mdpead) [inline R/feature_importance_permutation.R:136]: Could put this in the calculation on line 122 by just wrapping it in absolute function.

File: R/feature_importance_permutation.R [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+                                                  metric_kwarg))
+      scores_permuted_ii[kk] <- score_permuted # append to vector of scores
+    }
+
+    permuted_data[[var_ii]] <- original_values_ii # replace with original ordering
+
+    # create output for factor var_ii
+    out_df_ii <- data.frame(col_index=ii,
+                            factor=var_ii,
+                            importance_mean=score_standard - (scores_permuted_ii %>% mean()),
```
Comment (mdpead) [inline R/feature_importance_permutation.R:122]: I haven't used permutation importance before, but would expect it to be some relative measure from the `score_standard` here, say just divided through by `score_standard`. Or even coming out with a relative percentage of importance across all the features. Happy if this is the way it's usually calculated though.

File: R/feature_importance_permutation.R [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+                            importance_sd=scores_permuted_ii %>% sd(),
+                            stringsAsFactors = FALSE)
+
+    # append to existing output
+
+    if (ii==1){out_df <- out_df_ii}
+    else{out_df <- rbind(out_df, out_df_ii)}
```
Comment (mdpead) [inline R/feature_importance_permutation.R:129]: My R doesn't seem to like this section, but seems happier with if else is moved up to line 128...any ideas?

**Review by mdpead — COMMENTED** [review]
Still in the middle of it, but saving just in case!

File: R/feature_importance_permutation.R (hunk 14 above)
Comment (clementstross) [reply R/feature_importance_permutation.R:122]: I agree with you however I have copied the decision from
https://scikit-learn.org/stable/modules/generated/sklearn.inspection.permutation_importance.html#sklearn.inspection.permutation_importance

Comment (mdpead) [reply R/feature_importance_permutation.R:119]: Sweet, I'll bow to SKLearn.

**Review by mdpead — APPROVED** [review]
Looks good to me.

File: R/feature_importance_permutation.R (hunk 13 above)
Comment (CRSlade1) [reply R/feature_importance_permutation.R:133]: For anyone's future reference: Discussed but not true due to it being possible to have values worse than random.

File: R/feature_importance_permutation.R (hunk 10 above)
Comment (CRSlade1) [reply R/feature_importance_permutation.R:104]: For future reference: Discussed. Matt is right, it isn't needed but will be needed in 2D to isolate effects. Keeping here for continuity.

**Review by CRSlade1 — APPROVED** [review]
Thanks for making the changes asked for. Adding approval.
